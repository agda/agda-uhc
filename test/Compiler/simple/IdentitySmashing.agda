{-# OPTIONS -v treeless.opt:20 #-}

open import Agda.Builtin.Nat
open import Common.IO using (return)
open import Agda.Builtin.Unit

data List {a} (A : Set a) : Set a where
  []  : List A
  _∷_ : A → List A → List A

data Vec {a} (A : Set a) : Nat → Set a where
  [] : Vec A 0
  _∷_ : ∀ ..{n} → A → Vec A n → Vec A (suc n)

-- We can't handle smashing this yet. Different backends might compile
-- datatypes differently so we can't guarantee representational
-- compatibility. Possible solution: handle datatype compilation in
-- treeless as well and be explicit about which datatypes compile to the
-- same representation.
vecToList : ∀ {a} {A : Set a} ..{n} → Vec A n → List A
vecToList []       = []
vecToList (x ∷ xs) = x ∷ vecToList xs

data Fin : Nat → Set where
  zero : ∀ ..{n} → Fin (suc n)
  suc  : ∀ ..{n} → Fin n → Fin (suc n)

-- These should all compile to identity functions:

wk : ∀ ..{n} → Fin n → Fin (suc n)
wk  zero   = zero
wk (suc i) = suc (wk i)

wkN : ∀ ..{n m} → Fin n → Fin (n + m)
wkN  zero   = zero
wkN (suc i) = suc (wkN i)

wkN′ : ∀ m ..{n} → Fin n → Fin (m + n)
wkN′ zero i = i
wkN′ (suc m) i = wk (wkN′ m i)

vecPlusZero : ∀ {a} {A : Set a} ..{n} → Vec A n → Vec A (n + 0)
vecPlusZero [] = []
vecPlusZero (x ∷ xs) = x ∷ vecPlusZero xs

main = return tt
