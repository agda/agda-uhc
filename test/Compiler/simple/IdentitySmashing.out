EXECUTED_PROGRAM

ret > ExitSuccess
out > IdentitySmashing.vecToList =
out >   λ a b c d →
out >     case d of
out >       IdentitySmashing.Vec.[] → IdentitySmashing.List.[]
out >       IdentitySmashing.Vec._∷_ e f g →
out >         IdentitySmashing.List._∷_ f (IdentitySmashing.vecToList a _ _ g)
out > IdentitySmashing.wk = λ a b → b
out > IdentitySmashing.wkN = λ a b c → c
out > IdentitySmashing.wkN′ = λ a b c → c
out > IdentitySmashing.vecPlusZero = λ a b c d → d
out > IdentitySmashing.main = Common.IO.return () _ _
out >
